### PR TITLE
use parse to replay chatevents instead of pubnub

### PR DESCRIPTION
**Pull request by tonyjhuang**

### CODE REVIEW

File: cloud/main.js [hunk 1]
```diff
@@ -8,6 +8,8 @@
 */
 
 var _ = require('cloud/utils/underscore.js');
+var Moment = require('cloud/utils/moment.js');
```
Comment (deadlocked247) [inline cloud/main.js:11]: I usually make my imports const, but its not really a big deal rn

File: cloud/main.js [hunk 2]
```diff
@@ -84,28 +86,37 @@ function sendToFeedbackChannel(body, response) {
...
-    var requiredParams = ["aliasId", "subkey"];
+    var requiredParams = ["aliasId"];
     var params = request.params;
     checkMissingParams(params, requiredParams, response);
 
-    var count = params.count ? params.count : 9999;
-    var aliasId = params.aliasId;
+    var count = params.count || 20;
     var subkey = params.subkey;
+    var startTimeToken = params.startTimeToken ?
```
Comment (deadlocked247) [inline cloud/main.js:95]: yea generally dont use var in ES6 land, use let and const

File: cloud/main.js [hunk 3]
```diff
@@ -84,28 +86,37 @@ function sendToFeedbackChannel(body, response) {
...
+        new Moment(params.startTimeToken).toDate() : new Moment().toDate();
+    var endTimeToken = params.endTimeToken ? new Moment(params.endTimeToken).toDate() : undefined;
+    var aliasId = params.aliasId;
```
Comment (deadlocked247) [inline cloud/main.js:98]: `const { aliasId } = params;` is another way to do this
Plus you can do default values e.g. 
`const { aliasId = 'hey' } = params;`
this is destructing

File: cloud/main.js [hunk 4]
```diff
@@ -84,28 +86,37 @@ function sendToFeedbackChannel(body, response) {
...
-
-        Pubnub.replayChannel({subkey: subkey,
-                              channel: chatRoomId,
-                              startTimeToken: startTimeToken,
-                              endTimeToken: endTimeToken,
-                              count: count
-                             }).then(response.success, response.error);
-    }, response.error);
+    ChatEvent.getChatEvents(aliasId, startTimeToken, count, endTimeToken).then(function(chatEvents) {
+        var startTimeToken = chatEvents.length > 0 ?
```
Comment (deadlocked247) [inline cloud/main.js]: I would just use conditionals here, hard to read

File: cloud/main.js [hunk 5]
```diff
@@ -84,28 +86,37 @@ function sendToFeedbackChannel(body, response) {
...
+        function formatChatEvent(chatEvent) {
+            return {
+                "object": chatEvent,
+                "objectType": "ChatEvent"
+            }
+        }
+
+        return {
+            "startTimeToken": startTimeToken,
+            "endTimeToken": endTimeToken,
```
Comment (deadlocked247) [inline cloud/main.js:117]: unnecessary quotes

File: cloud/chatevent.js [hunk 6]
```diff
@@ -69,3 +70,24 @@ function getMostRecentForChatRoom(chatRoomId) {
...
 }
+
+function getChatEvents(aliasId, startDate, count, endDate) {
+    var aliasQuery = new Parse.Query("Alias");
+    return aliasQuery.get(aliasId).then(function(alias) {
+        var chatRoomId = alias.get("chatRoomId");
+        aliasQuery = new Parse.Query("Alias");
+        aliasQuery.equalTo("chatRoomId", chatRoomId);
+
+        var chatEventQuery = new Parse.Query("ChatEvent");
```
Comment (deadlocked247) [inline cloud/chatevent.js:81]: use single quotes everyywherreee

File: cloud/main.js [hunk 7]
```diff
@@ -84,28 +86,37 @@ function sendToFeedbackChannel(body, response) {
...
 
 Parse.Cloud.define("replayEvents", function(request, response) {
-    var requiredParams = ["aliasId", "subkey"];
+    var requiredParams = ["aliasId"];
     var params = request.params;
     checkMissingParams(params, requiredParams, response);
 
-    var count = params.count ? params.count : 9999;
-    var aliasId = params.aliasId;
+    var count = params.count || 20;
```
Comment (deadlocked247) [inline cloud/main.js:93]: using || is dangerous in this context, use default params like i said before

File: cloud/main.js [hunk 8]
```diff
@@ -84,28 +86,37 @@ function sendToFeedbackChannel(body, response) {
...
+    ChatEvent.getChatEvents(aliasId, startTimeToken, count, endTimeToken).then(function(chatEvents) {
+        var startTimeToken = chatEvents.length > 0 ?
+            new Moment(chatEvents[0].get("createdAt")).toISOString() : undefined;
+        var endTimeToken = chatEvents.length > 0 ?
+            new Moment(chatEvents[chatEvents.length - 1].get("createdAt")).toISOString() : undefined;
+
+        // For backwards compatibility.
+        function formatChatEvent(chatEvent) {
+            return {
+                "object": chatEvent,
```
Comment (deadlocked247) [inline cloud/main.js:110]: dont need quotes

File: cloud/main.js [hunk 9]
```diff
@@ -84,28 +86,37 @@ function sendToFeedbackChannel(body, response) {
...
-                              count: count
-                             }).then(response.success, response.error);
-    }, response.error);
+    ChatEvent.getChatEvents(aliasId, startTimeToken, count, endTimeToken).then(function(chatEvents) {
+        var startTimeToken = chatEvents.length > 0 ?
+            new Moment(chatEvents[0].get("createdAt")).toISOString() : undefined;
+        var endTimeToken = chatEvents.length > 0 ?
+            new Moment(chatEvents[chatEvents.length - 1].get("createdAt")).toISOString() : undefined;
+
+        // For backwards compatibility.
```
Comment (deadlocked247) [inline cloud/main.js:107]: use this format https://atom.io/packages/docblockr

File: cloud/main.js [hunk 10]
```diff
@@ -84,28 +86,37 @@ function sendToFeedbackChannel(body, response) {
...
-                              channel: chatRoomId,
-                              startTimeToken: startTimeToken,
-                              endTimeToken: endTimeToken,
-                              count: count
-                             }).then(response.success, response.error);
-    }, response.error);
+    ChatEvent.getChatEvents(aliasId, startTimeToken, count, endTimeToken).then(function(chatEvents) {
+        var startTimeToken = chatEvents.length > 0 ?
+            new Moment(chatEvents[0].get("createdAt")).toISOString() : undefined;
+        var endTimeToken = chatEvents.length > 0 ?
```
Comment (deadlocked247) [inline cloud/main.js]: making sure you have the right types is important when you do array logic

File: cloud/main.js [hunk 11]
```diff
@@ -84,28 +86,37 @@ function sendToFeedbackChannel(body, response) {
...
-        Pubnub.replayChannel({subkey: subkey,
-                              channel: chatRoomId,
-                              startTimeToken: startTimeToken,
-                              endTimeToken: endTimeToken,
-                              count: count
-                             }).then(response.success, response.error);
-    }, response.error);
+    ChatEvent.getChatEvents(aliasId, startTimeToken, count, endTimeToken).then(function(chatEvents) {
+        var startTimeToken = chatEvents.length > 0 ?
+            new Moment(chatEvents[0].get("createdAt")).toISOString() : undefined;
```
Comment (deadlocked247) [inline cloud/main.js]: you really shouldnt do ternary everywhere, especially when the last part is setting it to `undefined`

File: cloud/chatevent.js [hunk 12]
```diff
@@ -3,6 +3,7 @@ module.exports.createJoinPresence = createJoinPresence;
 module.exports.createLeavePresence = createLeavePresence;
 module.exports.createChangeRoomName = createChangeRoomName;
 module.exports.getMostRecentForChatRoom = getMostRecentForChatRoom;
+module.exports.getChatEvents = getChatEvents;
```
Comment (deadlocked247) [inline cloud/chatevent.js:6]: you can do this or 

```
export default = {
    createLeavePresence,
    createChangeRoomName,
   ....
}
```

and its usually at the bottom

File: cloud/main.js (hunk 7 above)
Comment (tonyjhuang) [reply cloud/main.js:93]: why is it dangerous?

File: cloud/main.js (hunk 4 above)
Comment (tonyjhuang) [reply cloud/main.js]: how would you do this? can you declare a var in an if statement and use it outside that block?

```
if(cond) {
  var x = 1
}
x + 2
```

File: cloud/main.js (hunk 10 above)
Comment (tonyjhuang) [reply cloud/main.js]: how would you do this then?

File: cloud/main.js (hunk 7 above)
Comment (deadlocked247) [reply cloud/main.js:93]: @tonyjhuang you could count as 0 which is falsely and will always give you 20

File: cloud/main.js (hunk 4 above)
Comment (deadlocked247) [reply cloud/main.js]: @tonyjhuang 
that works or just `var x;` above it

File: cloud/main.js (hunk 10 above)
Comment (deadlocked247) [reply cloud/main.js]: @tonyjhuang https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/isArray

File: cloud/main.js (hunk 7 above)
Comment (tonyjhuang) [reply cloud/main.js:93]: yeah thats true but we probably shouldn't let you set a count of 0 anyway.

File: cloud/main.js (hunk 10 above)
Comment (tonyjhuang) [reply cloud/main.js]: well im notchecking if it's an array, im checking if there are any elements

**Pull request merged**